models/src/main/scala/coop/rchain/models/rholangn/parmanager/Serialization.scala

```diff
@@ -1,164 +1,144 @@
-          write(pOpt.get)
-        } else write(false)
+  def write(pOpt: Option[RhoTypeN]): Eval[Unit] =
+    pOpt.map(write(true) *> write(_)).getOrElse(write(false))
 
-      private def write(kv: (ParN, ParN)): Unit = {
-        write(kv._1)
-        write(kv._2)
-      }
+  def writeTuplePar(kv: (RhoTypeN, RhoTypeN)): Eval[Unit] =
+    write(kv._1) *> write(kv._2)
 
-      private def writeInjection(injection: (String, ParN)): Unit = {
-        write(injection._1)
-        write(injection._2)
-      }
+  def writeTupleStringPar(kv: (String, RhoTypeN)): Eval[Unit] =
+    write(kv._1) *> write(kv._2)
 
-      private def writeSeq[T](seq: Seq[T], f: T => Unit): Unit = {
-        write(seq.size)
-        seq.foreach(f)
-      }
+  // Writes serialized value of a sequence
+  def writeSeq[T](seq: Seq[T], f: T => Eval[Unit]): Eval[Unit] =
+    write(seq.size) *> seq.traverse_(f)
+}
 
-      private def write(ps: Seq[RhoTypeN]): Unit           = writeSeq[RhoTypeN](ps, write)
-      private def writeStrings(strings: Seq[String]): Unit = writeSeq[String](strings, write)
-      private def writeKVPairs(kVPairs: Seq[(ParN, ParN)]): Unit =
-        writeSeq[(ParN, ParN)](kVPairs, write)
-      private def writeInjections(injections: Seq[(String, ParN)]): Unit =
-        writeSeq[(String, ParN)](injections, writeInjection)
+object Serialization {
+
+  // TODO: Properly handle errors
+  @SuppressWarnings(Array("org.wartremover.warts.TryPartial"))
+  def serializeToBytes(par: RhoTypeN): Eval[Array[Byte]] =
+    par.serializedSize.flatMap { serSize =>
+      Using(new ByteArrayOutputStream(serSize)) { baos =>
+        Serialization.serialize(par, baos).map { _ =>
+          baos.flush()
+          baos.toByteArray
+        }
+      }.get
+    }
 
-      private def write1ParOp(tag: Byte, p: ParN): Unit = {
-        write(tag)
-        write(p)
-      }
+  // TODO: Properly handle errors
+  @SuppressWarnings(Array("org.wartremover.warts.TryPartial"))
+  def deserializeFromBytes(bv: Array[Byte]): ParN =
+    Using(new ByteArrayInputStream(bv))(Serialization.deserialize(_).value).get
 
-      private def write2ParOp(tag: Byte, p1: ParN, p2: ParN): Unit = {
-        write(tag)
-        write(p1)
-        write(p2)
-      }
+  // TODO: Properly handle errors with return type (remove throw)
+  def serialize(par: RhoTypeN, output: OutputStream): Eval[Unit] = {
+    val cos         = CodedOutputStream.newInstance(output)
+    val protoWriter = new ProtobufPrimitiveWriter(cos)
 
-      @SuppressWarnings(Array("org.wartremover.warts.Throw"))
-      def write(p: RhoTypeN): Unit = p match {
+    // Serializer with recursive traversal of the whole object at once
+    lazy val serializer = new ProtobufRecWriter(protoWriter, writeRec)
 
-        /** Basic types */
-        case _: NilN.type => write(NIL)
+    // Serializer with recursive traversal using memoized values
+    // val serializer = new ProtobufRecWriter(protoWriter, _.serialized.flatMap(protoWriter.write))
```

Comment:
Serialization can work in two modes. To memoize serailized bytes on every object in the hierarchy or only on the top object. To automatically memoize on every object can take much more memory so now it's enabled serialization only on the called top object.